File: Prison Cells After N Days/Prison Cells After N Days.java
class Solution {
    // <bitmap, index>
    HashMap<Integer, Integer> map = new HashMap();
    public int[] prisonAfterNDays(int[] cells, int N) {
        // fast forwarding, record the state as a bitmap
        int cur_idx = 1;
        int cycle = -1;
        boolean forward = false;
        while(N!=0){
            int cur_state = state(cells);            
            if(!forward){
                if(map.containsKey(cur_state)){
                    cycle = cur_idx - map.get(cur_state);
                    //System.out.println(cycle);
                    N = N%cycle;
                    forward = true;
                    continue;
                }
                else{
                    map.put(cur_state, cur_idx++); 
                }
            }
            N--;   
            cells = next_day(cells);
        }

        return cells;
    }

    private int state(int[] cells){
        int state = 0;
        for(int cell: cells){
            state <<= 1;
            if(cell==1) state++;
        }
        return state;
    }

    private int[] next_day(int[] cells){
        int[] new_cells = new int[cells.length];
        new_cells[0] = 0;
        new_cells[cells.length-1]=0;
        for(int i=1; i<cells.length-1; i++){
            new_cells[i] = cells[i-1]==cells[i+1]?1:0;
        }
        return new_cells;
    }
}

File: Prison Cells After N Days/Prison Cells After N Days.py
class Solution:
    def prisonAfterNDays(self, cells: List[int], N: int) -> List[int]:
        self.d = collections.defaultdict(list)
        self.idx = 0
        self.loop = False
        self.cycle = 0
        while N>0:
            cells = self.nextDay(cells)
            N -=1
            if self.loop:
                N %= self.cycle
                break
        while N>0:
            cells = self.nextDay(cells)
            N -= 1
            
        return cells
    
    def nextDay(self, cells):
        nd = [0]*8
        for i in range(1,7):
            nd[i] = 1 if (cells[i-1]==cells[i+1]) else 0
        if not self.loop and tuple(nd) in self.d:
            self.loop = True
            self.cycle = self.idx - self.d[tuple(nd)]           
        elif not self.loop and tuple(nd) not in self.d:
            self.d[tuple(nd)] = self.idx
            self.idx+=1
        return nd